File: frontend/lib/api/subscription.ts
import apiClient from './client'

// 플랜 타입
export type PlanType = 'free' | 'basic' | 'pro' | 'business'

// 플랜 정보
export interface PlanInfo {
  type: PlanType
  name: string
  price_monthly: number
  price_yearly: number
  features: {
    keyword_search_daily: number
    blog_analysis_daily: number
    search_results_count: number
    history_days: number
    competitor_compare: number
    rank_alert: boolean
    excel_export: boolean
    api_access: boolean
    team_members: number
  }
}

// 구독 정보
export interface Subscription {
  user_id: number
  plan_type: PlanType
  plan_name: string
  billing_cycle: 'monthly' | 'yearly' | null
  status: 'active' | 'cancelled' | 'expired'
  started_at: string | null
  expires_at: string | null
  cancelled_at: string | null
  plan_limits: PlanInfo['features']
}

// 사용량 정보
export interface UsageInfo {
  date: string
  keyword_searches: {
    used: number
    limit: number
    remaining: number
  }
  blog_analyses: {
    used: number
    limit: number
    remaining: number
  }
  plan_type: PlanType
  plan_name: string
}

// 사용량 제한 체크 결과
export interface UsageLimitCheck {
  allowed: boolean
  used: number
  limit: number
  remaining: number
  plan: PlanType
  message?: string
  upgrade_message?: string
}

// 결제 준비 정보
export interface PaymentPrepareInfo {
  order_id: string
  order_name: string
  amount: number
  plan_type: PlanType
  billing_cycle: 'monthly' | 'yearly'
  customer_key: string
}

// 결제 내역
export interface PaymentHistory {
  id: number
  order_id: string
  amount: number
  status: string
  payment_method: string | null
  card_company: string | null
  card_number: string | null
  receipt_url: string | null
  paid_at: string | null
  created_at: string
}

// ============ 플랜 API ============

export async function getAllPlans(): Promise<PlanInfo[]> {
  const response = await apiClient.get('/api/subscription/plans')
  return response.data
}

export async function getPlanInfo(planType: PlanType): Promise<PlanInfo> {
  const response = await apiClient.get(`/api/subscription/plans/${planType}`)
  return response.data
}

// ============ 구독 API ============

// 사용량/구독 조회는 단순 PK 조회지만 backend 가 cron tick (naver API timeout
// 폭주 시 30s+ blocking) 동안 응답 못 함. scheduler 분리 후에도 worst-case 보호.
const LIGHT_READ_TIMEOUT_MS = 60000

// 일시적 timeout/네트워크 에러는 1회 재시도. 영구 4xx 는 즉시 throw.
async function lightRead<T>(path: string, params: Record<string, unknown>): Promise<T> {
  const attempt = async () => {
    const response = await apiClient.get(path, {
      params,
      timeout: LIGHT_READ_TIMEOUT_MS,
    })
    return response.data as T
  }
  try {
    return await attempt()
  } catch (e: any) {
    const code = e?.code
    const status = e?.response?.status
    // 4xx (auth/validation) 는 retry 무의미
    if (status && status >= 400 && status < 500) throw e
    // timeout / network / 5xx 만 1회 재시도 (250ms 백오프)
    if (code === 'ECONNABORTED' || code === 'ERR_NETWORK' || !status || status >= 500) {
      await new Promise(r => setTimeout(r, 250))
      return await attempt()
    }
    throw e
  }
}

export async function getMySubscription(userId: number | string): Promise<Subscription> {
  return lightRead<Subscription>('/api/subscription/me', { user_id: userId })
}

export async function upgradeSubscription(
  userId: number | string,
  planType: PlanType,
  billingCycle: 'monthly' | 'yearly' = 'monthly'
): Promise<{ success: boolean; message: string; subscription: Subscription }> {
  const response = await apiClient.post('/api/subscription/upgrade', {
    plan_type: planType,
    billing_cycle: billingCycle
  }, {
    params: { user_id: userId }
  })
  return response.data
}

export async function cancelSubscription(userId: number | string): Promise<{ success: boolean; message: string }> {
  const response = await apiClient.post('/api/subscription/cancel', null, {
    params: { user_id: userId }
  })
  return response.data
}

// ============ 사용량 API ============

export async function getUsage(userId: number | string): Promise<UsageInfo> {
  return lightRead<UsageInfo>('/api/subscription/usage', { user_id: userId })
}

export async function checkUsageLimit(
  userId: number | string,
  usageType: 'keyword_search' | 'blog_analysis'
): Promise<UsageLimitCheck> {
  const response = await apiClient.get('/api/subscription/usage/check', {
    params: { user_id: userId, usage_type: usageType }
  })
  return response.data
}

export async function incrementUsage(
  userId: number | string,
  usageType: 'keyword_search' | 'blog_analysis'
): Promise<{ success: boolean; usage?: any; used_extra_credit?: boolean }> {
  const response = await apiClient.post('/api/subscription/usage/increment', null, {
    params: { user_id: userId, usage_type: usageType }
  })
  return response.data
}

// ============ 결제 API ============

export async function preparePayment(
  userId: number | string,
  planType: PlanType,
  billingCycle: 'monthly' | 'yearly' = 'monthly'
): Promise<PaymentPrepareInfo> {
  const response = await apiClient.post('/api/payment/prepare', {
    plan_type: planType,
    billing_cycle: billingCycle
  }, {
    params: { user_id: userId }
  })
  return response.data
}

export async function confirmPayment(
  userId: number | string,
  paymentKey: string,
  orderId: string,
  amount: number
): Promise<{ success: boolean; message: string; payment: any }> {
  const response = await apiClient.post('/api/payment/confirm', {
    payment_key: paymentKey,
    order_id: orderId,
    amount: amount
  }, {
    params: { user_id: userId }
  })
  return response.data
}

export async function completeSubscriptionPayment(
  userId: number | string,
  paymentKey: string,
  orderId: string,
  planType: PlanType,
  billingCycle: 'monthly' | 'yearly'
): Promise<{ success: boolean; message: string; subscription: Subscription }> {
  const response = await apiClient.post('/api/payment/subscription/complete', null, {
    params: {
      user_id: userId,
      payment_key: paymentKey,
      order_id: orderId,
      plan_type: planType,
      billing_cycle: billingCycle
    }
  })
  return response.data
}

// 정기결제 등록 (빌링키 발급 + 첫 결제 + 구독 활성화)
export async function registerBilling(
  userId: number | string,
  customerKey: string,
  authKey: string,
  orderId: string,
  amount: number,
  planType: PlanType,
  billingCycle: 'monthly' | 'yearly'
): Promise<{ success: boolean; message: string; subscription: Subscription; payment: any }> {
  const response = await apiClient.post('/api/payment/billing/register', {
    customer_key: customerKey,
    auth_key: authKey,
    order_id: orderId,
    amount: amount,
    plan_type: planType,
    billing_cycle: billingCycle
  }, {
    params: { user_id: userId }
  })
  return response.data
}

export async function getPaymentHistory(
  userId: number | string,
  limit: number = 10
): Promise<{ payments: PaymentHistory[]; count: number }> {
  const response = await apiClient.get('/api/subscription/payments', {
    params: { user_id: userId, limit }
  })
  return response.data
}

// ============ 추가 크레딧 API ============

export async function getExtraCredits(userId: number | string): Promise<{ credits: any[] }> {
  const response = await apiClient.get('/api/subscription/credits', {
    params: { user_id: userId }
  })
  return response.data
}

export async function prepareCreditsPayment(
  userId: number | string,
  creditType: 'keyword' | 'analysis',
  amount: 100 | 500 | 1000
): Promise<{
  order_id: string
  order_name: string
  amount: number
  credit_type: string
  credit_amount: number
  customer_key: string
}> {
  const response = await apiClient.post('/api/payment/credits/prepare', null, {
    params: { user_id: userId, credit_type: creditType, amount }
  })
  return response.data
}

export async function completeCreditsPayment(
  userId: number | string,
  paymentKey: string,
  orderId: string,
  creditType: 'keyword' | 'analysis',
  creditAmount: number
): Promise<{ success: boolean; message: string; credit: any }> {
  const response = await apiClient.post('/api/payment/credits/complete', null, {
    params: {
      user_id: userId,
      payment_key: paymentKey,
      order_id: orderId,
      credit_type: creditType,
      credit_amount: creditAmount
    }
  })
  return response.data
}

File: flyio-backend/routers/subscription.py
"""
구독 관리 API 라우터
"""
from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
import logging

from database.subscription_db import (
    get_user_subscription,
    create_subscription,
    upgrade_subscription,
    cancel_subscription,
    get_today_usage,
    check_usage_limit,
    increment_usage,
    get_payment_history,
    get_extra_credits,
    add_extra_credits,
    use_extra_credit,
    PLAN_LIMITS,
    PlanType
)

logger = logging.getLogger(__name__)
router = APIRouter()


# ============ Pydantic 모델 ============

class SubscriptionResponse(BaseModel):
    user_id: int
    plan_type: str
    plan_name: str
    billing_cycle: Optional[str]
    status: str
    expires_at: Optional[str]
    plan_limits: dict


class UsageResponse(BaseModel):
    keyword_searches: int
    blog_analyses: int
    limits: dict
    plan_type: str


class UsageLimitCheck(BaseModel):
    allowed: bool
    used: int
    limit: int
    remaining: int
    plan: str


class UpgradeRequest(BaseModel):
    plan_type: str
    billing_cycle: str = "monthly"


class PlanInfo(BaseModel):
    type: str
    name: str
    price_monthly: int
    price_yearly: int
    features: dict


# ============ 플랜 정보 API ============

@router.get("/plans", response_model=List[PlanInfo])
async def get_all_plans():
    """모든 구독 플랜 정보 조회"""
    plans = []
    for plan_type, limits in PLAN_LIMITS.items():
        plans.append({
            "type": plan_type.value,
            "name": limits["name"],
            "price_monthly": limits["price_monthly"],
            "price_yearly": limits["price_yearly"],
            "features": {
                "keyword_search_daily": limits["keyword_search_daily"],
                "blog_analysis_daily": limits["blog_analysis_daily"],
                "search_results_count": limits["search_results_count"],
                "history_days": limits["history_days"],
                "competitor_compare": limits["competitor_compare"],
                "rank_alert": limits["rank_alert"],
                "excel_export": limits["excel_export"],
                "api_access": limits["api_access"],
                "team_members": limits["team_members"],
            }
        })
    return plans


@router.get("/plans/{plan_type}")
async def get_plan_info(plan_type: str):
    """특정 플랜 정보 조회"""
    try:
        plan = PlanType(plan_type)
        limits = PLAN_LIMITS[plan]
        return {
            "type": plan.value,
            "name": limits["name"],
            "price_monthly": limits["price_monthly"],
            "price_yearly": limits["price_yearly"],
            "features": limits
        }
    except ValueError:
        raise HTTPException(status_code=404, detail="플랜을 찾을 수 없습니다")


# ============ 구독 관리 API ============

@router.get("/me")
async def get_my_subscription(user_id: int = Query(..., description="사용자 ID")):
    """내 구독 정보 조회"""
    subscription = get_user_subscription(user_id)

    if not subscription:
        # 구독이 없으면 무료 플랜으로 자동 생성
        subscription = create_subscription(user_id, "free")

    return {
        "user_id": subscription["user_id"],
        "plan_type": subscription["plan_type"],
        "plan_name": PLAN_LIMITS[PlanType(subscription["plan_type"])]["name"],
        "billing_cycle": subscription.get("billing_cycle"),
        "status": subscription["status"],
        "started_at": subscription.get("started_at"),
        "expires_at": subscription.get("expires_at"),
        "cancelled_at": subscription.get("cancelled_at"),
        "plan_limits": subscription["plan_limits"]
    }


@router.post("/upgrade")
async def upgrade_plan(
    request: UpgradeRequest,
    user_id: int = Query(..., description="사용자 ID")
):
    """구독 업그레이드 (결제 완료 후 호출)"""
    try:
        plan = PlanType(request.plan_type)
    except ValueError:
        raise HTTPException(status_code=400, detail="유효하지 않은 플랜입니다")

    subscription = upgrade_subscription(
        user_id=user_id,
        plan_type=request.plan_type,
        billing_cycle=request.billing_cycle
    )

    logger.info(f"User {user_id} upgraded to {request.plan_type} ({request.billing_cycle})")

    return {
        "success": True,
        "message": f"{PLAN_LIMITS[plan]['name']} 플랜으로 업그레이드되었습니다",
        "subscription": subscription
    }


@router.post("/cancel")
async def cancel_plan(user_id: int = Query(..., description="사용자 ID")):
    """구독 취소 (만료일까지 유지)"""
    success = cancel_subscription(user_id)

    if not success:
        raise HTTPException(status_code=404, detail="구독을 찾을 수 없습니다")

    return {
        "success": True,
        "message": "구독이 취소되었습니다. 만료일까지 서비스를 이용하실 수 있습니다."
    }


# ============ 사용량 API ============

@router.get("/usage")
async def get_usage(user_id: int = Query(..., description="사용자 ID")):
    """오늘 사용량 조회"""
    subscription = get_user_subscription(user_id)
    if not subscription:
        subscription = create_subscription(user_id, "free")

    usage = get_today_usage(user_id)
    limits = subscription["plan_limits"]

    return {
        "date": usage["date"],
        "keyword_searches": {
            "used": usage["keyword_searches"],
            "limit": limits["keyword_search_daily"],
            "remaining": max(0, limits["keyword_search_daily"] - usage["keyword_searches"])
                if limits["keyword_search_daily"] != -1 else -1
        },
        "blog_analyses": {
            "used": usage["blog_analyses"],
            "limit": limits["blog_analysis_daily"],
            "remaining": max(0, limits["blog_analysis_daily"] - usage["blog_analyses"])
                if limits["blog_analysis_daily"] != -1 else -1
        },
        "plan_type": subscription["plan_type"],
        "plan_name": limits["name"]
    }


@router.get("/usage/check")
async def check_limit(
    user_id: int = Query(..., description="사용자 ID"),
    usage_type: str = Query(..., description="사용 유형 (keyword_search, blog_analysis)")
):
    """사용량 제한 확인"""
    if usage_type not in ["keyword_search", "blog_analysis"]:
        raise HTTPException(status_code=400, detail="유효하지 않은 사용 유형입니다")

    result = check_usage_limit(user_id, usage_type)

    if not result["allowed"]:
        plan_limits = PLAN_LIMITS[PlanType(result["plan"])]
        return {
            **result,
            "message": f"일일 {usage_type} 한도에 도달했습니다. (무료: {plan_limits['keyword_search_daily']}회)",
            "upgrade_message": "더 많은 검색을 원하시면 베이직 플랜으로 업그레이드하세요!"
        }

    return result


@router.post("/usage/increment")
async def record_usage(
    user_id: int = Query(..., description="사용자 ID"),
    usage_type: str = Query(..., description="사용 유형 (keyword_search, blog_analysis)")
):
    """사용량 기록 (내부용)"""
    if usage_type not in ["keyword_search", "blog_analysis"]:
        raise HTTPException(status_code=400, detail="유효하지 않은 사용 유형입니다")

    # 먼저 제한 확인
    limit_check = check_usage_limit(user_id, usage_type)

    if not limit_check["allowed"]:
        # 추가 크레딧 확인
        credit_type = "keyword" if usage_type == "keyword_search" else "analysis"
        if use_extra_credit(user_id, credit_type):
            return {
                "success": True,
                "used_extra_credit": True,
                "message": "추가 크레딧을 사용했습니다"
            }

        raise HTTPException(
            status_code=429,
            detail={
                "message": "일일 한도에 도달했습니다",
                "limit": limit_check["limit"],
                "used": limit_check["used"],
                "plan": limit_check["plan"]
            }
        )

    usage = increment_usage(user_id, usage_type)
    return {
        "success": True,
        "usage": usage
    }


# ============ 결제 내역 API ============

@router.get("/payments")
async def get_payments(
    user_id: int = Query(..., description="사용자 ID"),
    limit: int = Query(10, description="조회 개수")
):
    """결제 내역 조회"""
    payments = get_payment_history(user_id, limit)
    return {
        "payments": payments,
        "count": len(payments)
    }


# ============ 추가 크레딧 API ============

@router.get("/credits")
async def get_credits(user_id: int = Query(..., description="사용자 ID")):
    """추가 크레딧 잔여량 조회"""
    credits = get_extra_credits(user_id)
    return {
        "credits": credits
    }


@router.post("/credits/purchase")
async def purchase_credits(
    user_id: int = Query(..., description="사용자 ID"),
    credit_type: str = Query(..., description="크레딧 유형 (keyword, analysis)"),
    amount: int = Query(..., description="구매 수량")
):
    """추가 크레딧 구매 (결제 완료 후 호출)"""
    if credit_type not in ["keyword", "analysis"]:
        raise HTTPException(status_code=400, detail="유효하지 않은 크레딧 유형입니다")

    if amount not in [100, 500, 1000]:
        raise HTTPException(status_code=400, detail="유효하지 않은 구매 수량입니다")

    credit = add_extra_credits(user_id, credit_type, amount)

    return {
        "success": True,
        "message": f"{amount}개의 {credit_type} 크레딧이 추가되었습니다",
        "credit": credit
    }
